api/eth_types.go

```diff
@@ -32,19 +49,41 @@ func (e *EthInt) UnmarshalJSON(b []byte) error {
 		return err
 	}
 	eint := EthInt(parsedInt)
-	e = &eint
+	*e = eint
 	return nil
 }
 
 type EthBigInt big.Int
 
+var (
+	EthBigIntZero = EthBigInt{Int: big.Zero().Int}
+)
+
 func (e EthBigInt) MarshalJSON() ([]byte, error) {
 	if e.Int == nil {
 		return json.Marshal("0x0")
 	}
 	return json.Marshal(fmt.Sprintf("0x%x", e.Int))
 }
 
+func (e *EthBigInt) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	replaced := strings.Replace(s, "0x", "", -1)
+	if len(replaced)%2 == 1 {
+		replaced = "0" + replaced
+	}
+
+	i := new(mathbig.Int)
+	i.SetString(replaced, 16)
+
+	*e = EthBigInt(big.NewFromGo(i))
+	return nil
```

Comment:
Any reason why not to use the `encoding/hex` package? https://pkg.go.dev/encoding/hex

Reply:
It's likely that the underlying string is a big int, so it's easier to parse the string directly using the standard big.Int library.